src/Policies.cpp

```diff
@@ -118,30 +118,40 @@ std::vector<Policy *> PolicyManager::load(const YAML::Node &policy_yaml)
                 throw PolicyException(fmt::format("invalid input config for tap '{}': {}", tap_name, e.what()));
             }
         }
+        // TODO separate config and filter. for now, they merge
+        tap_filter.config_merge(tap_config);
 
-        // Create Policy
-        auto policy = std::make_unique<Policy>(policy_name, tap);
-        // if and only if policy succeeds, we will return this in result set
-        Policy *policy_ptr = policy.get();
+        std::string input_stream_module_name;
+        tap_filter.config_hash(input_stream_module_name);
+        input_stream_module_name.insert(0, tap->name() + "-");
 
-        // Instantiate stream from tap
         std::unique_ptr<InputStream> input_stream;
-        std::string input_stream_module_name;
-        try {
-            spdlog::get("visor")->info("policy [{}]: instantiating Tap: {}", policy_name, tap_name);
-            // TODO separate config and filter. for now, they merge
-            tap_filter.config_merge(tap_config);
-            input_stream = tap->instantiate(policy.get(), &tap_filter);
-            // ensure tap input type matches policy input tap
-            if (input_node["input_type"].as<std::string>() != tap->input_plugin()->plugin()) {
-                throw PolicyException(fmt::format("input_type for policy specified tap '{}' doesn't match tap's defined input type: {}/{}", tap_name, input_node["input_type"].as<std::string>(), tap->input_plugin()->plugin()));
+        InputStream *input_ptr;
+        if (_registry->input_manager()->module_exists(input_stream_module_name)) {
+            spdlog::get("visor")->info("policy [{}]: input stream already exists. reusing: {}", policy_name, input_stream_module_name);
+            std::unique_lock<std::shared_mutex> input_lock;
+            auto result_input = _registry->input_manager()->module_get_locked(input_stream_module_name);
+            input_ptr = result_input.module;
+            input_lock = std::move(result_input.lock);
```

Comment:
this lock will be released immediately because input_lock is defined in this block - I think it needs to move out of the block to where input_stream is defined

Reply:
lock outside the scope will cause deadlock, probably we should not lock at all. what do you think?

Reply:
we definitely need a few test cases with removals

Reply:
we do have test cases to cover that. the deadlock is in the new test case:
```
/home/lparente/git/pktvisor/src/tests/test_policies.cpp:636: FAILED:
1:   REQUIRE_NOTHROW( registry.policy_manager()->load(config_file2["visor"]["policies"]) )
1: due to unexpected exception with message:
1:   Resource deadlock avoided
```

Reply:
When I lock the input mutex in the context

Reply:
So, just to clarify. The deadlock I've mentioned was running Unit tests also mentioned above.

I found the race point. The possible deadlock was that `input_lock` is the mutex from input manager map. Latter on in the Policy Load code there is
```
         if (!_registry->input_manager()->module_exists(input_stream_module_name)) {
                _registry->input_manager()->module_add(std::move(input_stream));
            }
```

module_exists also uses the same mutex to interate over inputStreams map